### ISSUE TITLE
Crash when rendering animation and using BMesh node within blender 2.8

### ISSUE THREAD

**Issue by javismiles** [nortikin/sverchok#2417]
- Blender 2.8
- Using a number of nodes that end on a BMesh node 
- There are a number of frames where a shape changes through the frames being rendered by BMesh
- when rendering single frames all is good
- when rendering an animation, blender will crash after 1 or 2 frames are rendered

* tested deleting the material of the shape, crash still happens
* tested disabling the LIVE option of the BMesh node, Crash does Not happen then, but the shape, of course remains static and does not change between frames. Re-enabling the LIVE option in BMesh brings back the crash when rendering the animation
* tested with multiple resolutions, very small ones as well, different rendering parameters, very low quality settings etc, and the crash still happens when LIVE is enabled in BMesh and the shape changes every frame and an animation is rendered

Thank you very much to @zeffii and Sverchok team for the support with this issue and others

![alt text](https://github.com/javismiles/torch/blob/master/sam2.jpg?raw=true "example")

**Comment by enzyme69** [comment]
Crashing while rendering with SV processing nodes is quite common. I recommend you to try baking out as Alembic sequence, hopefully that can resolve the crashing issue. There are of course many case where it is not possible to bake out mesh or vertex color... but try.

**Comment by zeffii** [comment]
well, @javismiles  - let's see what we can achieve.  Historically it's not been a passion of mine to fix animation issues, but I might be using it in the future too and that becomes incentive.

first, let's establish a few ground truths / states.

- can we make sverchok/blender crash with a most minimal node setup that ends in a BMesh Node
    - something stupid simple like a `sphere node`  that goes directly to a BMeshViewer, and where only the radius of the Sphere is keyframed.

- In the case of the Vertex / Topology staying the same, BMeshViewer allows also an even faster mode (available from the side panel) than merely updates the Vertex locations. I think it's called 'fixed vertex count' -- you'll see it.

if that also crashes, then I think we might have to implement a ' do not update nodetree ' (which the BMeshViewer would control internally) in the part of the BMeshViewer code that handles the Mesh update.

**Comment by javismiles** [comment]
> Crashing while rendering with SV processing nodes is quite common. I recommend you to try baking out as Alembic sequence, hopefully that can resolve the crashing issue. There are of course many case where it is not possible to bake out mesh or vertex color... but try.

Good day Jimmy and thank you :) correct me if I'm wrong but if I have an animation of 1000 positions of the shape, baking the 1000 frames to produce 1000 separated baked shapes and then showing them or hiding them in turn doesn't sound like a great solution to me or even feasible to be honest, but correct me please if I'm wrong :)  I really wish that we can correct this issue, looks like @zeffii will look at this and I am here to help in all that I can because I really want this to be fixed :D

**Comment by javismiles** [comment]
> well, @javismiles - let's see what we can achieve. Historically it's not been a passion of mine to fix animation issues, but I might be using it in the future too and that becomes incentive.
> 
> first, let's establish a few ground truths / states.
> 
> * can we make sverchok/blender crash with a most minimal node setup that ends in a BMesh Node
>   
>   * something stupid simple like a `sphere node`  that goes directly to a BMeshViewer, and where only the radius of the Sphere is keyframed.
> * In the case of the Vertex / Topology staying the same, BMeshViewer allows also an even faster mode (available from the side panel) than merely updates the Vertex locations. I think it's called 'fixed vertex count' -- you'll see it.
> 
> if that also crashes, then I think we might have to implement a ' do not update nodetree ' (which the BMeshViewer would control internally) in the part of the BMeshViewer code that handles the Mesh update.

@zeffii a million thanks because I really really need to render animations with this and I am here to help you in all that I can to fix this, I am grateful if all together we can try to fix this :D 

Got your message and I will today try that very simple example with BMesh and tell you if that also crashes, I'm on it ;)

**Comment by javismiles** [comment]
@zeffii I have news, useful news
I made the simplest possible thing and it crashes! Look

- A single sphere connected to a BMesh Node
- 50 frames, no changes, just the sphere static
- Render animation: Crash, what I notice is that sometimes crash happens after frame 1, other times it happened after 7 or 8 frames, its variable, but always crash anyway

Obviously there is a big issue here, and now we have the simples possible setup to find out where in the code the bug is :)

Here is the .blend file:
https://github.com/javismiles/torch/blob/master/animate1.blend

**Comment by javismiles** [comment]
@zeffii , and like always
- Disabling the LIVE option of BMesh stops the crashes, no crash then
- Re-Enabling the LIVE option reactivates the Crash after x frames, sometimes after 1 frame, other times after 5 or 7 frames

that "fixed vertex count" mode you mentioned I couldn't find it sorry, maybe tell me more about where to find it and I can also try that

in any case the simples possible setup crashes even without keyframing anything
https://github.com/javismiles/torch/blob/master/animate1.blend

so hopefully that can give us clues as to where in the code the issue may be :)

**Comment by zeffii** [comment]
the panel that appears when you press N in nodetree, has a sverchok tab.. and a Node tab.  Some nodes have additional properties that don't really fit in the nodeview node UI, and for other reasons we stick those  slightly less immediately accessible ( on purpose )

![image](https://user-images.githubusercontent.com/619340/57760205-74cd6f00-76fb-11e9-8945-0febd1bce7ae.png)

cool. i'll update Blender and synchronize with the latest version of this b28_prelease_master and do some experiments.

**Comment by zeffii** [comment]
also... The `Live`  , option is what tells the node to be _active_ ,  it's like muting the node when unset.

**Comment by javismiles** [comment]
zeffii  cool, following your instructions I now found the side panel and the Fixed Vertex Count option and guess what

- with fixed vertex count activated: tried twice, once it didnt crash during the 50 frames, another time it crashed after 30 frames

- with fixed vertex count deactivated: it always crashes, but sometimes it crashes after 13 frames, other times after 7 frames, other times after 2 frames

so far we can say that fixed vertex count activated reduces the chance of crash although it still happened to me one of the 2 times I tried

**Comment by javismiles** [comment]
@zeffii 
actually it doesnt help much, I tried again with fixed vertex count activated and this time it did crash after only 2 frames

So I would say that it doesnt really help the fixed vertex count change

**Comment by zeffii** [comment]
I think....untested..... this section would need to be enclosed by a 'freeze the nodetree from the effects of the blender depsgraph updates feature`

```python
from sverchok.utils.context_managers import hard_freeze

with hard_freeze(self) as _: 
    <insert snippet
```
https://github.com/nortikin/sverchok/blob/b28_prelease_master/nodes/viz/viewer_bmesh.py#L258-L299

**Comment by zeffii** [comment]
ok @javismiles  ,   we have other nodes that would likely suffer the same issue, i'll check this out at my earliest convenience.

**Comment by javismiles** [comment]
> ok @javismiles , we have other nodes that would likely suffer the same issue, i'll check this out at my earliest convenience.

@zeffii thank you very much, crossing fingers that we can fix it, I'm here to help testing more if you need it :)

**Comment by zeffii** [comment]
maybe we should write a decorator for the process function -  to enter and exit a freezed state (even if exceptions are encountered in the process evaluation)

```python
@freeze_nodetree_for_safety
def process(self):
    ...
```

**Comment by javismiles** [comment]
@zeffii what I find quite interesting is that playing the animation inside blender, Eveee, etc, it all plays perfectly, smoothly and with no crashes.  Crashes happen when "rendering" the animation. So it's the process of rendering that causes the issue. 
Interesting because I thought that with Eevee, playing the animation in the viewport and rendering the animation would pretty much do the same stuff, except writing the file in the second case, but obviously there must be more differences, as playing the animation doesn't crash but rendering it does crash.

**Comment by javismiles** [comment]
> @zeffii what I find quite interesting is that playing the animation inside blender, Eveee, etc, it all plays perfectly, smoothly and with no crashes. Crashes happen when "rendering" the animation. So it's the process of rendering that causes the issue.
> Interesting because I thought that with Eevee, playing the animation in the viewport and rendering the animation would pretty much do the same stuff, except writing the file in the second case, but obviously there must be more differences, as playing the animation doesn't crash but rendering it does crash.

**Comment by zeffii** [comment]
now, i've just added a freeze feature as described above,  please confirm that that also does not fix the crash ( my blender will also crash to the desktop )  after frame 8

**Comment by vicdoval** [comment]
@javismiles by the moment you an use the script in  https://github.com/nortikin/sverchok/issues/2390#issuecomment-492654519 to bake the animation to shapekeys and turn off sverchok when rendering.
Just select the object produced by the bmesh viewer and run the script.

**Comment by javismiles** [comment]
> now, i've just added a freeze feature as described above, please confirm that that also does not fix the crash ( my blender will also crash to the desktop ) after frame 8

@zeffii to test this, do I have to download the latest b28_master branch zip and reinstall it in blender, correct?

**Comment by javismiles** [comment]
> @javismiles by the moment you can use the script in [#2390 (comment)](https://github.com/nortikin/sverchok/issues/2390#issuecomment-492654519) to bake the animation to shapekeys and turn off sverchok when rendering.
> Just select the object produced by the bmesh viewer and run the script.

thank you @vicdoval , let me see if I understand this well, so once I have the animation ready, outputting to a BMesh, I run that script in blender, and then how do I turn sverchok off, do u mean the turn the LIVE option of BMesh to off for example? thank you again

**Comment by javismiles** [comment]
@vicdoval yeah this works, using that script to bake the animation to shapekeys and disabling then BMesh works to at least for now export the animation in that way, thank you very much, thats a nice temporary solution while we try to fix the issue itself, let's see if what @zeffii is trying with the freeze feature solves the issue in the meantime, thank you both so much

**Comment by zeffii** [comment]
> @zeffii to test this, do I have to download the latest b28_master branch zip and reinstall it in blender, correct?

i think you just press update now in b28_pre_release, I hardcoded the location of the zip it would download/  failing that, then dl manually and reinstall, yes

**Comment by javismiles** [comment]
@vicdoval  @zeffii  question, whats the easiest way to add a renderable wireframe on top of the sverchok shape? thank you very much

**Comment by javismiles** [comment]
> > @zeffii to test this, do I have to download the latest b28_master branch zip and reinstall it in blender, correct?
> 
> i think you just press update now in b28_pre_release, I hardcoded the location of the zip it would download/ failing that, then dl manually and reinstall, yes

got you, I go try now

**Comment by javismiles** [comment]
@zeffii just did it, redownloaded and reinstall and crash still happens yeah, same crash indeed

**Comment by javismiles** [comment]
@zeffii  yeah re-confirmed, crash still happens with latest

**Comment by javismiles** [comment]
@zefii , with the more complex scenario of the shape animation i got the same crashes after just 1 frame, and with the super simple sphere I get crash after 8 frames also,

**Comment by zeffii** [comment]
time to go into debug mode then. no ETA

**Comment by zeffii** [comment]
or maybe something like  https://blender.stackexchange.com/a/110717/47

**Comment by vicdoval** [comment]
With https://github.com/nortikin/sverchok/issues/2390#issuecomment-492712076 i just rendered 250 frames of animated bmesh (without the last sverchok freeze update)

**Comment by javismiles** [comment]
@zeffii, thank you, so:

- crossing fingers for the debug, no ETA I know

- I just tried that other script that renders through code with 3 second pause between frames, it seems to work, really hacky as blender goes totally blank while rendering but seems to work,  with one issue, that the background comes solid black, and should be transparent as its set to transparent, mmm, any idea
why it's not coming transparent same as when rendering a single frame inside blender?

nice hack while we wait to see if fix is possible, but im trying to see if I can make the background be transparent as it should be

thank u ;)

**Comment by javismiles** [comment]
> With [#2390 (comment)](https://github.com/nortikin/sverchok/issues/2390#issuecomment-492712076) i just rendered 250 frames of animated bmesh (without the last sverchok freeze update)

hey Im gonna try this now

**Comment by javismiles** [comment]
@zeffii @vicdoval here feedback

the code just posted by vicdoval at #2390 and a previous one of zeffii (https://blender.stackexchange.com/a/110717/47) both work but the latest is faster and works better, this works yes except for the one puzzling thing that the background gets rendered solid black in the PNG when it should be transparent, its set to transparent in blender and rendering separated frames in blender renders background transparent, any idea why the script is not rendering the background transparent?

**Comment by javismiles** [comment]
in any case this is good news because I really don't care if I render using the render button in blender or running a script, as long as the animation render works I'm happy, just need to see now why the background is getting rendered solid black when it should be transparent

**Comment by nortikin** [comment]
@javismiles also this https://github.com/nortikin/sverchok/blob/master/node_scripts/SNLite_templates/shape_key_bake.py

**Comment by javismiles** [comment]
> @javismiles also this https://github.com/nortikin/sverchok/blob/master/node_scripts/SNLite_templates/shape_key_bake.py
> but not work in 2.8...

thank you @nortikin but yeah blender 2.8 is essential and the other script shared by @vicdoval works really well yeah

**Comment by zeffii** [comment]
ok. later will try to make a mode specifically for when blender is rendering.

i think an event handler for `pre_render`, should be switched on, and both `depsgraph_update` and `on_frame_change` would be ended early

**Comment by zeffii** [comment]
note to self..

or...  if  scene.is_animating, then ignore depsgraph events, only listen to frame change. (post)

**Comment by zeffii** [comment]
there's now ` layer.update() ` which might be needed in frame_change event handler , instead of `scene.update()`

**Comment by javismiles** [comment]
friends, quick one about Bmesh, do you have any idea why with a shape created by Bmesh (baked or not baked) when viewed in the shader area of Blender 2.8, I can see the shading well both from inside and from outside of it (inner and outer faces), but when viewing it in the main viewport and when rendering it, only the inner space/faces get the shading/lighting applied but the outer ones, the outside area does not? picture attached, left side is view from shading area, right one is view from main viewport/render, etc, thank u for any tips ;)

![alt text](https://raw.githubusercontent.com/javismiles/torch/master/shader.jpg "example")

**Comment by javismiles** [comment]
@zeffii, quick doubt, about the Plane MK2 node, it works great, but the resulting mesh ends up kind of not sharp enough, is there any way to increase the sharpness of the result? increasing the number of points surely helps but only to a point and I cannot have millions of points so I wonder if there are other tricks to increase the sharpness of the resulting mesh while still keeping it smooth, thank you very much

**Comment by zeffii** [comment]
i think you're referring to 'faceting' ,  you could use the N panel option to set ' smooth ' . this will interpolate the surface normals.

Plane node just providese topology, you pass the number of verts x,y plus their location.  make the resolution as high as you computer can handle. then set the smooth attribute.

**Comment by zeffii** [comment]
using the smooth option will let you get away with a lot less geometry ..mostly.

**Comment by javismiles** [comment]
@zeffii thank you very much, question, this smooth option is where exactly? I click N and the side panel appears but I can't see a smooth option ;)  I hope that will work better than the other thing I tried, I applied a Subdivision modifier to the mesh and this made it much smoother but also kind of blurry, smooth yes but not sharp, what I'm looking for is smooth + sharp, let me know if possible where this smooth option is and I will test that one, thank you ;)

**Comment by javismiles** [comment]
@zeffii mmm I see the settings of the nodes in the N panel but cant see a smooth option,   as for the resolution I have to put the one of the csv I generated, I have already tested with csvs up to 250000 points, but still need some trick to make the mesh sharper and smoother, lets see if i find this smooth option and test it ;)

**Comment by javismiles** [comment]
@zeffii now I see it, "Smooth Shade" option, its in the BMesh node, ok, I tried it, yes this makes a very similar result to using a Subdivision modifier, it smoothens the mesh a lot, which is great, but at the same time the result gets kind of more blurry rather than sharp, what I'm looking for mostly is sharpness (with a degree of smoothness), any tips? :)

**Comment by zeffii** [comment]
if you are working with a fixed number of verts, you could set the 'fixed vert count' attribute to True, and also add a SubDiv modifier to the Mesh generated by Sverchok.  There's no way to do this from the BMeshViewer node (for reasons.... )

that might work.

else, you might want to show an example of the renders ( 1. flat, 2. smoothshade, 3. subdiv...) , it might be something unrelated to SV.  (ie how well do you know the render system )

**Comment by javismiles** [comment]
@zeffii thank you very much, look I tried it all, I show you this picture where you can see the original 300x300 (90000 points), then the version with smooth shade on the node (great smoothing but it looks too blurry), then the version with fixed vert count and subdiv modifier (this one is better than the previous one but still too blurry),  I also show in the picture my shading nodes, I'm using Eevee as my renderer and those are my shading nodes, let me know if now that you see this pic if there are any other tips to make the result smooth enough yet as sharp as possible so that it doesn't look blurry :D thank you :D 
![alt text](https://raw.githubusercontent.com/javismiles/torch/master/imgs/sverchok-sharp-smooth.jpg "example")

**Comment by zeffii** [comment]
can you render any sharp material on that mesh? this is using a gradient ColorRamp afterall, ..it appears to me that that will have a continous colour transition. 

for experiment make a more rigid colour ramp which results in very clear bands, just to see how the renderer behaves on the threshold of fullback to fullwhite. 

I really, sadly, have had almost no time with Eevee and don't know if there are settings you could adjust to allow more resampling.  What happens in Cycles (with lots of sampling ) ?

**Comment by javismiles** [comment]
thank you @zeffii I'm gonna check what you say and report back, didn't test cycles much but cycles is not really viable for what I'm doing as renders would take too long

**Comment by javismiles** [comment]
@zeffii thank you again, so yes, if I make the color ramp use the constant mode to make super sharp transitions between the colors, the render has those super sharp transitions between the colors, but those are just the colors. As we know, we always have two aspects, tone and chroma. The color ramp influences the chroma, not the tones, so yes I can make the color transitions be sharp but that does not help me at all with what I need, which is sharpness in the tonal distribution.

I also tried cycles now but impossible, that would take forever, gotta use Eevee. The renderer can clearly render sharp, for example I can generate a sharp wireframe with no issues. Problem here basically is that the smoothing treatments, both the sverchok node one and the subdiv modifier one apply too much smoothing and blur the resulting model making it lose sharpness

I guess will have to research more and see, thank you again @zeffii ;)

**Comment by zeffii** [comment]
maybe use an emission shader ?  instead of diffuse bdsf

**Comment by javismiles** [comment]
> maybe use an emission shader ? instead of diffuse bdsf

@zeffii just tried emission shader, unfortunately that literally sends tonal info out and just makes a smooth colored surface all around ;) gotta try something else ;)

**Comment by zeffii** [comment]
recently several Blender dev commits addressed render issues in 2.8.  I retested this here, and rendered 50 frames without crashing. Might be worth checking again.

**Comment by zeffii** [comment]
note there are some console warnings about calling `Scene.update()`. if I change that to the latest API i think it's something like `layers.update()`, it might break again. In which case the whole call might be redundant during animation.

**Comment by zeffii** [comment]
i've made some commits, which might be useful.

FYI, rendering a really simple frame (like a sphere increasing and shrinking in size at standard render size, takes about 4 minutes in Eevee ( on my machine ) .   using Workbench renderer is much faster, and definitely doesnt crash. Cycles also works, but is ofcourse slow as a slug :)

**Comment by zeffii** [comment]
there are render settings in Eevee to gather more samples ( in the properties panel ) , the default for render is 64 .. maybe up that?

**Comment by zeffii** [comment]
related:  https://developer.blender.org/T68505   maybe

**Comment by javismiles** [comment]
Hi friends, @zeffii  , it´s been a while, I wanted to ask you if with the latest blender 2.8 and the latest sverchok, if situation has improved regarding rendering the animations

at the moment I am still using this script to do the animations:

`
import bpy 
file_path = 'xxxxxxxxxx'

start_frame = 1
end_frame =95

for act_frame in range(start_frame, end_frame + 1):
    bpy.context.scene.frame_current = act_frame
    bpy.ops.node.sverchok_update_all()
    bpy.context.scene.render.filepath = file_path + str(act_frame).zfill(4) +'.png'
    bpy.ops.render.render(write_still=True)
    print(act_frame)

print ('done')
`
is it now with latest versions possible to render animations without crashing? thank you for any info ;)

**Comment by zeffii** [comment]
well, freezing during animation turns out was being experienced also by non-sverchok users.

- some people have reported that enabling "freeze ui" while rendering increases stability

but I haven't modified sverchok to try to rectify this issue since we last discussed it. I can only suggest that you do the experiment yourself.  (maybe simplify your scene and try again as a less resource-hugging test)

**Comment by javismiles** [comment]
@zeffii @vicdoval  quick one, is it possible with sverchok that if I have for example 100 points that define a curve, can I give these 100 points in 3D space to sverchok and that sverchok draws the curve through the 100 points? if so, using which modules? thank you so much :)

**Comment by vicdoval** [comment]
In viz menu try some of this
 ![image](https://user-images.githubusercontent.com/10011941/67125638-96280f00-f1f5-11e9-823c-0b2a6a32103e.png)
You may also need UV connect. :) and maybe a 'Vector Sort' or a 'KDT closest path' if vectors are not sorted

**Comment by javismiles** [comment]
thank you very much @vicdoval , I managed to create the line by doing this
![alt text](https://raw.githubusercontent.com/javismiles/torch/blob/master/imgs/solve.jpg "example")

connecting the line module to the edges, so now the line appears, great, but I need to give it volume or something to be able to render it and give it color and see it, what would you use from this state? thank you very much :)

**Comment by javismiles** [comment]
@vicdoval oh yes, changing to curveviewer works, now the line appears and I can render it, awesome ;) but how can I change its width and color? any idea? thank you ;)

**Comment by zeffii** [comment]
colour you set via a material.

**Comment by javismiles** [comment]
@zeffii yes I used it but its not changing, let me show you, a sec

**Comment by zeffii** [comment]
the default `+` material expects Cycles renderer, i'll adapt this tomorrow to check which renderer is used before spawning material nodes..

**Comment by zeffii** [comment]
ok show me :)

**Comment by javismiles** [comment]
see here
![alt text](https://github.com/javismiles/torch/blob/master/imgs/here.jpg?raw=true "example")

by the way why is the curves module red, like it has an error, but its working well, thank you ;)

**Comment by javismiles** [comment]
@zeffi and how can I make that the points are connected via soft curves rather than sharp corners? thank you ;)

**Comment by javismiles** [comment]
@zeffii and even with cycles still rendering without color so far

**Comment by zeffii** [comment]
there are probably several features that I've neglected to implement in the viewers.

why is it red, no idea, show the command prompt / terminal window output.

play around with the settings, but i would use Polyline Viewer. it requires no edges.

**Comment by zeffii** [comment]
![image](https://user-images.githubusercontent.com/619340/67127835-6a5b5800-f1fa-11e9-9b8b-b1cb46fce946.png)

**Comment by javismiles** [comment]
@zeffii  wow you are right, I just used polyline viewer and works without edges, awesome, but its still white and sharp, how can I change color and how can I make soft transitions between points? ;)

**Comment by javismiles** [comment]
@zeffii  yes, bspiline makes it soft, great! now the color left only ;)

**Comment by javismiles** [comment]
@zeffii color works also! all works now ;)

**Comment by javismiles** [comment]
@zeffii only one thing, the bspline make it all too soft, like makes it all a soft curve but that is actually away from many of the points, is there a way to make this soft interpolation more or less pronounced? awesome job by the way sverchok is simply the best of the best ;)

**Comment by zeffii** [comment]
![image](https://user-images.githubusercontent.com/619340/67128121-0dac6d00-f1fb-11e9-82c2-a496ad7c685f.png)

**Comment by zeffii** [comment]
ok, if you want the points to stick closer to the originals then you might have to stick an Interpolation node infront of the points ..  or a Smooth Lines node..

![image](https://user-images.githubusercontent.com/619340/67128443-f0c46980-f1fb-11e9-96ca-3cf4cecc4bc3.png)

**Comment by zeffii** [comment]
if you can draw a set of representative 2d vectors, and draw the desired kind of line through them, then it would be possible to approach answering that with more certainty.

**Comment by zeffii** [comment]
as in  .. you draw with grease pencil / mspaint / whatever , as an illustration :)

**Comment by javismiles** [comment]
@zeffii that´s so great, by the way I dont know how you managed with just the random vector MK2 to do such a beautiful spiral shape, I just used it and I get nothing like that

as for my shape, some collaborators are gonna give me 300 points that will define a bezier curve, and I will have to do this, but I still dont have their data ;) but it will be quite  a simple curve, and the curve will change every frame, so I will have to do the same I did with the mesh months ago, but with the curve, it will change every frame

now that I see that I can do the curve, I gotta see how to combine both things, the mesh changing every frame and the curve changing every frame, I wonder if I should just duplicate twice the process mmm

**Comment by zeffii** [comment]
> ust the random vector MK2 to do such a beautiful spiral shape, I just used it and I get nothing like that

heheh, i jogged through the seed value until something pleasing came up.

**Comment by javismiles** [comment]
@zeffii very clever ;) hey in which menu is smooth-lines module? my search doesnt work in blender 2.8 no idea why thank you

**Comment by zeffii** [comment]
it's `shift A` now  or `alt+space`

**Comment by zeffii** [comment]
in the CAD section

**Comment by javismiles** [comment]
@zeffii found it yeah, added it but it doesnt seem to help to get closer to the points mm

**Comment by zeffii** [comment]
it really depends what you want, i'd need to understand that before answering.  a lot of possible : )

![image](https://user-images.githubusercontent.com/619340/67129122-0470cf80-f1fe-11e9-827d-d7549459a94f.png)

**Comment by zeffii** [comment]
that uses the `Vector Interpolation` node

**Comment by javismiles** [comment]
@zeffii  wow your vector interpolation node looks different to mine, yours has more options,
and why do you need the bottom part? the UV correction and VD experimental

my scenario is a math related process and through each frame of the process I receive 300 points that make a curve, and I just have to visualize them changing frame by frame ;)

**Comment by zeffii** [comment]
the viewer draw is just to visualize the incoming stream of vectors, un-adulterated.

**Comment by javismiles** [comment]
@zeffi yes got it, now I got the advanced one, looks like yours, and yes this seems to make it better than the smooth lines

**Comment by zeffii** [comment]
and you can add two interpolators.. and use a Lerp node to interpolate between them .. till it looks good..

**Comment by javismiles** [comment]
@zeffii yes this is kind of the solution awesome ;)

now in your opinion, remember what I did months ago, now in parallel I gotta do the same but with the line that changes through the frames, so I guess I will make a parallel execnode chain to do that and I have both updating both mesh and line every frame

**Comment by zeffii** [comment]
![image](https://user-images.githubusercontent.com/619340/67129712-aa710980-f1ff-11e9-87ba-052952fb6a4b.png)

**Comment by zeffii** [comment]
try it and see,  or a Script Node.   anything that can read current frame and output a set of verts appropriate to that frame.

**Comment by zeffii** [comment]
i don't have adequate hardware here to give tips on rendering, i'll defer to @vicdoval  :)

**Comment by javismiles** [comment]
@zeffii @vicdoval thank you both for your great support and for creating this incredible marvel ;) I will try it and if I get stuck will return to ask for advice, have a great day ;)

**Comment by javismiles** [comment]
@zeffii , @vicdoval , quick one, when having an animation of around 500 frames, and each frame with many thousands of points, it gets extremely slow to do anything in blender, move camera or anything, is there any trick to avoid this, it seems as if the system is trying to process the entire animation every time I do anything in any of the frames, thank you ;)

**Comment by Durman** [comment]
You can just switch off update during animation on N panel.
![2019-11-05_08-17-51](https://user-images.githubusercontent.com/28003269/68178321-0cbe5e00-ffa5-11e9-8443-54e9a59586b9.png)

**Comment by javismiles** [comment]
thank you @Durman , yes I would love to do that but I cannot find that option anywhere, I´m on blender 2.8, where is that option, can´t find it in the N panel, thank you very much for your help

**Comment by javismiles** [comment]
@Durman I found it, yes if I disable all of them, that one, also the P on the right, and also the one on the left, then things go faster, still not as fast as one would want (I´m on a very powerful machine, latest i9, 32GB Ram, nvidia gtx card, etc), any other tricks on top of that one to speed things up when moving between frames? thanks very much ;)   also what´s the difference between those 3 blue icons? thank you ;)

**Comment by javismiles** [comment]
@Durman issue is that yes moving between frames now has improved, but doing anything, like moving lights or anything on the interface, often lags for a long time, this would never happen with 1 frame but happens with 500 of these frames, so again its as if somehow the fact that there are 500 frames continues weighting on the time it takes; memory is not the problem, I have 32gb of ram, and this is taking according to blender around 14gb of ram in total

**Comment by zeffii** [comment]
out of curiosity; what happens if you restrict the frame range from `1...500` to just the frames you are working on... f.ex  `[244..255]`

**Comment by javismiles** [comment]
@zeffii just tried, restricted range to 150 to 165, for example, then  I did same test, I have 3 lights, so I just click on each of them on the top right panel to switch between selecting each of them, and it seems to behave pretty similar with the restricted range as with the full 500 frames range

**Comment by zeffii** [comment]
I don't see why the "frame range" or "range of frames that are keyframed" would have this particular effect on sverchok. Sverchok's update system may need an overhaul. We could also add a master  "do not recalculate" toggle, for during scene edits..

 There are things we don't know about `b28` until users such as yourself try to make sverchok do things beyond what we ourselves do.

**Comment by javismiles** [comment]
@zeffii what is interesting is that if in the top right panel I deactivate visualizing the generated object both in viewport and in render, the lags still happen

**Comment by zeffii** [comment]
for example, this handler is called whenever the scene updates  (scene updates are caused by anything... even sneezing)..

```
@persistent
def sv_main_handler(scene):
    """
    Main Sverchok handler for updating node tree upon editor changes
    """
    for ng in sverchok_trees():
        # print("Scene handler looking at tree {}".format(ng.name))
        if ng.has_changed:
            print('depsgraph_update_pre called - ng.has_changed -> ')
            # print('sv_main_handler')
            # print("Edit detected in {}".format(ng.name))
            ng.process()

```
here `ng.has_changed` may still be defaulting to true where we expect it to be false.

**Comment by javismiles** [comment]
@zeffii what I can tell you for sure is that when I load 1 single frame at 300x300 (90K points), no issue, good speed; when I put 500 of those frames then we got this massive lag problems, and again the ram memory is not the problem, I got 32 gigs of ram, this takes around 14, and the most powerful i9 processor with a good nvidia gtx card, mmm, any trick I could experiment with to try to bypass what you just mentioned? ;)

**Comment by zeffii** [comment]
>  visualizing the generated object both in viewport and in render

this doesn't affect the node's calculations, only how Blender chooses to display them in the scene or not.  The `UPD` button would however..

**Comment by javismiles** [comment]
@zeffii after hiding and disabling literally everything in the viewport, lights, the object, etc, the lags still happen, so yeah it looks to be some internal processing going on, I Wonder if I Can try anything to try to change that mmm

**Comment by javismiles** [comment]
> for example, this handler is called whenever the scene updates (scene updates are caused by anything... even sneezing)..
> 
> ```
> @persistent
> def sv_main_handler(scene):
>     """
>     Main Sverchok handler for updating node tree upon editor changes
>     """
>     for ng in sverchok_trees():
>         # print("Scene handler looking at tree {}".format(ng.name))
>         if ng.has_changed:
>             print('depsgraph_update_pre called - ng.has_changed -> ')
>             # print('sv_main_handler')
>             # print("Edit detected in {}".format(ng.name))
>             ng.process()
> ```
> 
> here `ng.has_changed` may still be defaulting to true where we expect it to be false.

I guess ideally whenever anything in the editor updates sverchok wouldnt do anything, that way this problem would be solved I guess

**Comment by javismiles** [comment]
@zeffii I´ve been having this issue since the beginning but it happens proportional to the number of frames, and until now I was dealing with 200 or so, but now I need to do 500 and it becomes almost impossible to use it because of these massive lags, 
let me know if I can try to make any change in the code or anywhere to prevent this processing every time anything is touched in editor, etc, thank you ;)

**Comment by zeffii** [comment]
yeah, check back in 20 minutes

**Comment by javismiles** [comment]
@zeffii I always thought these lags were strange because I use a very powerful machine, and every time you touch anything in the editor, select something or etc, suddenly this lag of x number of seconds, and proportional to the number of frames I have in the animation, so yeah Im sure there must be something we can try in the code to disable this processing every time something is touched in the editor, ok Zeffi I check back at that time, thank you for your help :)

**Comment by zeffii** [comment]
yesyes. there's a property of the nodetree "sv_process" which i didn't include in the update handler. When i do the "P" button would do what it's supposed to do.. (i think.. )

**Comment by javismiles** [comment]
@zeffii ok great zeffii, Im here to try whatever you tell me to try, I really need to fix this somehow , thank you thank you :D

**Comment by zeffii** [comment]
you'll need to update, ive just pushed it.

**Comment by javismiles** [comment]
@zeffii  so you made a change, so I should update the plugin in my blender right, ok could you remind me the best place to download the new version from? thank you very much

**Comment by zeffii** [comment]
https://github.com/nortikin/sverchok/pull/2299/commits/c96a5497dd91166a392c8b4f1362e7f05d34a72b

you can press update.. i think from the toolbar.

**Comment by javismiles** [comment]
@zeffii great I go try now and let u know what happens ;)

**Comment by zeffii** [comment]
now it's not automatic, Sverchok doesn't know when to process and when not to process, so you have to use that P toggle (P for process)  set it to off when you are moving stuff about, and switch it back on when you want to render / see updates.

**Comment by javismiles** [comment]
@zeffii  thats great as long as I can control it myself, so you mean that with the new update, when disabling the P , all should be immediate when using the editor, and I can activate back the P whenever I want to render or when I want to analyze the look of the different frames; but when I want to build the animation, lights etc I could disable P and all should go smooth

I hope it will work ;)

I just did a test with 1 single frame and all goes so fast, I need the same when I have 500

ok I tried to update from toolbar but gives error (cannot get archive from the internet), any tips?

thank you ;)

**Comment by javismiles** [comment]
@zeffii  by the way it says that latest update is b4c188e by Ilya Portnov, maybe I have an older version,
maybe best to reinstall from the beginning your latest, where can I get it fully thank you ;)

**Comment by javismiles** [comment]
@zeffii downloading the sverchok-master.zip from the master branch, should be your latest one, right? I can then try to reinstall the entire plugin in blender

**Comment by zeffii** [comment]
noooooooooooooooooooooo

**Comment by zeffii** [comment]
the zip is not "master" .  
option 1 : https://github.com/nortikin/sverchok/archive/b28_prelease_master.zip   , install manually
option 2 : run a small script, then update (the script will correct the path to the zip and download automatically )

**Comment by javismiles** [comment]
@zeffii  my current version is this one:
sverchok-b28_prelease_master

**Comment by javismiles** [comment]
@zeffii  great, then I can just download that zip, and literally replace the current files with those ones

**Comment by javismiles** [comment]
@zeffii I got the zip down, can I just literally go to the place in my computer where the files are and literally delete them and replace them for the contents of this zip, right?

**Comment by zeffii** [comment]
option 2 :  run this in the Blender text editor, the update sverchok
```
    import sverchok
    from sverchok.utils import sv_prefs

    sv_prefs.set_vals(
        dload_archive_path="https://github.com/nortikin/sverchok/archive/",
        dload_archive_name="b28_prelease_master"
    )
```

**Comment by zeffii** [comment]
if that fails.. then yeah, reinstall sverchok .. see the thread https://github.com/nortikin/sverchok/issues/2326

**Comment by javismiles** [comment]
@zeffii that script failed, gave error, so I will just replace the files and try ;)

**Comment by zeffii** [comment]
what error ? :)

**Comment by javismiles** [comment]
@zeffii I see the instructions to install sverchok, prob is that I already have this version installed in a very different path, here:
C:\Users\xxxx\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons

so what do I do, do I just first remove the plugin and then do what that page says,
or simply go and in that same folder where it is now, delete the files and replace them with the new ones ;)

the error only says to check the error in the system console which I havent got on at the moment  ;)

**Comment by zeffii** [comment]
> delete the files and replace them with the new ones ;)

yeah

> error in the system console   

you can look at the the `info` View, it should tell you something.

**Comment by javismiles** [comment]
now the script works, it was a matter of the tabs yeah

**Comment by javismiles** [comment]
@zeffii so the script works, but well, whats easiest now, replacing the files is super easy, so maybe thats still the easiest

**Comment by zeffii** [comment]
just press update :)

**Comment by zeffii** [comment]
if that fails.. go ahead and manually replace the files

**Comment by zeffii** [comment]
the update should be relatively fast for you :)

**Comment by javismiles** [comment]
@zeffii did it, I clicked update and apparently succeeded, any way to check for sure that Im on the latest version? ;)

**Comment by javismiles** [comment]
@zeffii clicking on "check for updates" now says that no new commits exist, so I guess im on the latest one ;) ok now Im gonna do the key test ;) will load the 500 frame one, and check that P icon, to see if it solves the prob ;) let u know in a minute

**Comment by javismiles** [comment]
@zeffii to be clear, of the 3 icons in blue, the only one I should touch now is the "P" one right?, disable it to do stuff with editor, and enable it to render and to see the shape change in the frames

**Comment by zeffii** [comment]
yeah, with P off,  sverchok does no processing when that handler is called.   perhaps if there is still processing happening, another handler is called.. maybe the undo/redo handler...  it will be a matter of sticking a few print statements around the place.. (ie .. in the handlers... you can do this yourself to test... )  

but now I must take over other business..

**Comment by javismiles** [comment]
@zeffii thank you Zeffii, yeh I just tested and there is no change, same issue happens as before, a pity, thank you very much for trying to help, I will try to look for that part of code and see if I can detect anything

**Comment by javismiles** [comment]
@zeffii oh oh, now we have a problem, even with P activated, the frames are not updating anymore Zeffii..... oooops, they never change now, with or without P, I guess I will have to try to revert to previous version

**Comment by javismiles** [comment]
@zeffii  ohhh I got tons of errors in the sverchok elements "undefined socket" something, thats why they dont change, all the sverchok elements have errors now

**Comment by zeffii** [comment]
sigh....  you've been using and oooold 2.8 version.

**Comment by javismiles** [comment]
@zeffii yes thats what it seems, because I went back to the old version (i did a backup fortunately), and now it all works well again....    ;) so mmmm, what do I do, so you did a lot of changes in the meantime I guess, the version I use is from months ago, remember when we communicated around May 2019

**Comment by javismiles** [comment]
@zeffii in any case I must say that the lag was still happening with the new version, even with all the errors in the sverchok nodes; in any case, why all these errors mmm, you changed a lot of things, does that mean that with the new version, compared with the May one, I would have to reconstruct all my nodes again, mmm, that´s a prob, I have so many blender files now using the May version mmm

**Comment by javismiles** [comment]
@zeffii well I guess I can keep using the May version and just manually change things in the file where you made the changes mmm

**Comment by zeffii** [comment]
Let me explain, because this is as annoying for me as it might be for you. A while back we decided to rename our socket "identifiers" to avoid conflicts with other node systems. (namespace pollution...) 

it might be possible to convert a layout onthe fly..
```
def rename(sockets):
    for socket in sockets:
        if socket.bl_idname .... blablablablab

def node_tree_name(tree_name):
    import bpy
    for node in bpy.data.node_groups[tree_name].nodes:
        rename(node.inputs)
        rename(node.outputs)
```

**Comment by javismiles** [comment]
@zeffii  went back to new version to show you the errors I get 
![alt text](https://github.com/javismiles/torch/blob/master/imgs/err.jpg?raw=true"example")

**Comment by javismiles** [comment]
@zeffii mmm, I see zeffii, mmm, well, maybe instead of doing that, looks messy, I could just stick to the may version for now, and just do manual changes on the code as needed for now

**Comment by zeffii** [comment]
no. please hold on.  stick with the new one for 5minutes..

**Comment by javismiles** [comment]
@zeffii I can stick to the May version for now, and go to the file core/handlers.py, and do the change you did and then we see if we find other ways to do other changes

**Comment by javismiles** [comment]
@zeffii sure sure, I keep it today in the new version, if you have any easy ideas to fix it of course I will try them

**Comment by javismiles** [comment]
@zeffii will be here for the next 2 hours, then gotta go out and back 5 hours later

**Comment by javismiles** [comment]
@zeffii but yes if there is a solution should be a pretty easy one, because imagine, I have now tons of blender files using the may version, if I have to fix each one of them... better be an easy fix :D :D

**Comment by zeffii** [comment]
i think this renames the sockets... 
-----
# if it fails.. 
------
# do not press save

```python
socket_renames = {
    # to old
    "SvStringsSocket": "StringsSocket",
    "SvVerticesSocket": "VerticesSocket",
    "SvMatrixSocket": "MatrixSocket",
    # to new
    "StringsSocket": "SvStringsSocket", 
    "VerticesSocket": "SvVerticesSocket",
    "MatrixSocket": "SvMatrixSocket"
}


def rename(sockets):
    for socket in sockets:
        socked_found = socket_renames.get(socket.bl_idname)
        if socket_found:
            socket.bl_idname = socket_found

def node_tree_name(tree_name):
    import bpy
    for node in bpy.data.node_groups[tree_name].nodes:
        rename(node.inputs)
        rename(node.outputs)
```

**Comment by javismiles** [comment]
@zeffii  tried it , errors:

2019-11-05 12:25:14,878 [ERROR] sverchok.core.update_system: Node Text in+ had exception: 'bpy_prop_collection[key]: key "f454.csv" not found'
Traceback (most recent call last):
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\update_system.py", line 325, in do_update_general
    node.process()
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\nodes\text\text_in_mk2.py", line 257, in process
    self.update_csv()
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\nodes\text\text_in_mk2.py", line 281, in update_csv
    self.reload_csv()
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\nodes\text\text_in_mk2.py", line 305, in reload_csv
    self.load_csv_data()
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\nodes\text\text_in_mk2.py", line 329, in load_csv_data
    f = io.StringIO(bpy.data.texts[self.text].as_string())
KeyError: 'bpy_prop_collection[key]: key "f454.csv" not found'
2019-11-05 12:25:14,892 [ERROR] sverchok.core.update_system: Node Object ID Selector had exception: 'NodeSocket' object has no attribute 'sv_set'
Traceback (most recent call last):
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\update_system.py", line 325, in do_update_general
    node.process()
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\nodes\object_nodes\get_asset_properties.py", line 202, in process
    output_socket.sv_set(data_list[:])
AttributeError: 'NodeSocket' object has no attribute 'sv_set'
get_and_set_gl_scale_info called from sv_post_load
getting gl scale params
set render_scale to: 2.5
set render_location_xy_multiplier to: 2.5
stethoscope update holdout (not a problem)
Loaded ListItem2Node. 0 nodes are left unregisted.
Traceback (most recent call last):
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\handlers.py", line 167, in sv_post_load
    upgrade_nodes.upgrade_nodes(ng)
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\upgrade_nodes.py", line 170, in upgrade_nodes
    if socket and not socket.prop_name:
AttributeError: 'NodeSocket' object has no attribute 'prop_name'
2019-11-05 12:25:15,378 [ERROR] sverchok.core.update_system: Node Text in+ had exception: 'NodeSocket' object has no attribute 'sv_set'
Traceback (most recent call last):
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\update_system.py", line 325, in do_update_general
    node.process()
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\nodes\text\text_in_mk2.py", line 257, in process
    self.update_csv()
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\nodes\text\text_in_mk2.py", line 298, in update_csv
    self.outputs[name].sv_set([csv_data[name]])
AttributeError: 'NodeSocket' object has no attribute 'sv_set'
2019-11-05 12:25:15,387 [ERROR] sverchok.core.update_system: Node Object ID Selector had exception: 'NodeSocket' object has no attribute 'sv_set'
Traceback (most recent call last):
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\update_system.py", line 325, in do_update_general
    node.process()
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\nodes\object_nodes\get_asset_properties.py", line 202, in process
    output_socket.sv_set(data_list[:])
AttributeError: 'NodeSocket' object has no attribute 'sv_set'
Traceback (most recent call last):
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\sockets.py", line 276, in draw
    layout.label(text=text + '. ' + SvGetSocketInfo(self))
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\socket_data.py", line 57, in SvGetSocketInfo
    s_id = other.socket_id
AttributeError: 'NodeSocket' object has no attribute 'socket_id'

location: <unknown location>:-1

location: <unknown location>:-1
Traceback (most recent call last):
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\sockets.py", line 276, in draw
    layout.label(text=text + '. ' + SvGetSocketInfo(self))
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\socket_data.py", line 57, in SvGetSocketInfo
    s_id = other.socket_id
AttributeError: 'NodeSocket' object has no attribute 'socket_id'

location: <unknown location>:-1

location: <unknown location>:-1
Traceback (most recent call last):
  File "C:\MEDIA\LOSSlandscapes\BlenderBases\updatesverchok.blend\Text", line 25, in <module>
  File "C:\MEDIA\LOSSlandscapes\BlenderBases\updatesverchok.blend\Text", line 21, in rename_all_sockets_in_node_tree
KeyError: 'bpy_prop_collection[key]: key "your nodetree name" not found'
Error: Python script failed, check the message in the system console

**Comment by zeffii** [comment]
turn off all updates.  (unpress the P)

**Comment by zeffii** [comment]
then run the script..

**Comment by zeffii** [comment]
> KeyError: 'bpy_prop_collection[key]: key "your nodetree name" not found'

that should be ..the name of your node tree...  maybe `NodeTree.003`  whatever

**Comment by zeffii** [comment]
![Uploading image.png…]()

**Comment by javismiles** [comment]
@zeffii did that, name of my node is CSV_TO_MESH

errors 

Traceback (most recent call last):
  File "xxxxTextText", line 25, in <module>
  File "xxxxTextText", line 22, in rename_all_sockets_in_node_tree
  File "xxxxText", line 16, in rename
NameError: name 'socket_found' is not defined
Error: Python script failed, check the message in the system console

**Comment by javismiles** [comment]
location: <unknown location>:-1
Traceback (most recent call last):
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\sockets.py", line 276, in draw
    layout.label(text=text + '. ' + SvGetSocketInfo(self))
  File "C:\Users\ideam\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\sverchok-b28_prelease_master\core\socket_data.py", line 57, in SvGetSocketInfo
    s_id = other.socket_id
AttributeError: 'NodeSocket' object has no attribute 'socket_id'

**Comment by zeffii** [comment]
i made a typo in that code snippet.. also this handles NodeReroutes..
```python
socket_renames = {
    # to old
    "SvStringsSocket": "StringsSocket",
    "SvVerticesSocket": "VerticesSocket",
    "SvMatrixSocket": "MatrixSocket",
    # to new
    "StringsSocket": "SvStringsSocket", 
    "VerticesSocket": "SvVerticesSocket",
    "MatrixSocket": "SvMatrixSocket"
}


def rename(sockets):
    for socket in sockets:
        socket_found = socket_renames.get(socket.bl_idname)
        if socket_found:
            socket.bl_idname = socket_found

def rename_all_sockets_in_node_tree(tree_name):
    import bpy
    for node in bpy.data.node_groups[tree_name].nodes:
        
        if node.bl_idname == 'NodeReroute:
            continue

        rename(node.inputs)
        rename(node.outputs)

rename_all_sockets_in_node_tree("your nodetree name")
```

**Comment by javismiles** [comment]
File "xxx\Text", line 23
    if node.bl_idname == 'NodeReroute:
                                     ^
SyntaxError: EOL while scanning string literal

location: <unknown location>:-1

location: <unknown location>:-1
Error:   File "xxx\Text", line 23
    if node.bl_idname == 'NodeReroute:
                                     ^
SyntaxError: EOL while scanning string literal

location: <unknown location>:-1

Error: Python script failed, check the message in the system console

**Comment by zeffii** [comment]
hmmm there was a typo in that too... it's fixed in the previous post..

**Comment by zeffii** [comment]
but even then , the script isn't doing what i expected..  (i've found an old .blend... )

**Comment by javismiles** [comment]
@zeffii mmm, its a pity mmm, would be great if it could be fixed with a little script yeah

**Comment by zeffii** [comment]
yeah, i thought this would work but sadly, also new socket classes would need to be registered..
```python
socket_renames = {
    "SvStringsSocket": "StringsSocket",
    "SvVerticesSocket": "VerticesSocket",
    "SvMatrixSocket": "MatrixSocket",
    "StringsSocket": "SvStringsSocket", 
    "VerticesSocket": "SvVerticesSocket",
    "MatrixSocket": "SvMatrixSocket"
}

import bpy

def rename(sockets):
    for socket in sockets:
        socket_found = socket_renames.get(socket.bl_idname)
        if socket_found:
            print(f'renaming bl_idname {socket.bl_idname}->{socket_found}')
            socket.bl_idname = socket_found

def rename_all_sockets_in_node_tree(tree_name):
    for node in bpy.data.node_groups[tree_name].nodes:
        
        if node.bl_idname == 'NodeReroute':
            continue

        print(node)
        rename(node.inputs)
        rename(node.outputs)

rename_all_sockets_in_node_tree("NodeTree")
```

**Comment by zeffii** [comment]
hmm... maybe it's possible..

**Comment by zeffii** [comment]
yeah, it's trickier than I thought...

**Comment by zeffii** [comment]
something like this would be needed in the `pre_blend_load` event handler, and be executed if sverchok is told to "register old sockets" , for cases like this..
```python
import sverchok
import bpy

from sverchok.core.sockets import SvVerticesSocket, SvMatrixSocket, SvStringsSocket

class VerticesSocket(SvVerticesSocket):
    ...

class MatrixSocket(SvMatrixSocket):
    ...

class StringsSocket(SvStringsSocket):
    ...

classes = [VerticesSocket, MatrixSocket, StringsSocket]
register, unregister = bpy.utils.register_classes_factory(classes)

if __name__ == '__main__':
    register()
```

**Comment by javismiles** [comment]
@zeffii yes I guess it´s difficult, so from tomorrow I guess will go back to May version, and there I can manually change lines in the code, the one you suggested and others here and there, let´s see, and in the meantime maybe if one day you find an easy fix for this will try it for sure ;)

**Comment by javismiles** [comment]
@zeffii and about the lag with many frames, I guess will check that code see if there are other events that I can experiment with mm

**Comment by javismiles** [comment]
> @zeffii yes I guess it´s difficult, so from tomorrow I guess will go back to May version, and there I can manually change lines in the code, the one you suggested and others here and there, let´s see, and in the meantime maybe if one day you find an easy fix for this will try it for sure ;)

I 

> something like this would be needed in the `pre_blend_load` event handler, and be executed if sverchok is told to "register old sockets" , for cases like this..
> 
> ```python
> import sverchok
> import bpy
> 
> from sverchok.core.sockets import SvVerticesSocket, SvMatrixSocket, SvStringsSocket
> 
> class VerticesSocket(SvVerticesSocket):
>     ...
> 
> class MatrixSocket(SvMatrixSocket):
>     ...
> 
> class StringsSocket(SvStringsSocket):
>     ...
> 
> classes = [VerticesSocket, MatrixSocket, StringsSocket]
> register, unregister = bpy.utils.register_classes_factory(classes)
> 
> if __name__ == '__main__':
>     register()
> ```

I see @zeffii , yeh I guess its not so easy to do this

**Comment by zeffii** [comment]
i might have time for this later today.

**Comment by zeffii** [comment]
ok. that's not going to happen.   you might have more luck exporting the blend as a .json and running a script over the json to rename all of the mentioned socketypes sockets, prior to loading the json.

I really want to avoid adding code to the loader. breaking blends was anticipated.

**Comment by javismiles** [comment]
good day @zeffii, I get you, for now I continue with the older version, another question if I may, at the moment I am loading csv files with the coordinates, I may have to also work with .stl files (https://en.wikipedia.org/wiki/STL_(file_format)) , do you know if there is a way to load .stl format directly into sverchok in the same way that I use csvs now? thank you ;)

**Comment by javismiles** [comment]
hey @zeffii , is there anyway with sverchok to represent a gaussian distribution?  thank you ;)

**Comment by zeffii** [comment]
the random number generator has a "normal" distribution mode, which is the same thing .

**Comment by zeffii** [comment]
also the statistics node might be useful, i think it does buckets

**Comment by javismiles** [comment]
@zeffi thats cool, yes I was gonna generate the x,y,z in python and then just import it as csv and use the typical viewer mesh node and solved, but I may try what you suggest, although if I need distributions different to the standard normal, if I need different mean and variance then I guess best is to generate the coordinates in python and import them into sverchok as usual to make the mesh

other than that, what about the stl thing, I may have to also work with .stl files (https://en.wikipedia.org/wiki/STL_(file_format)) , do you know if there is a way to load .stl format directly into sverchok in the same way that I use csvs now? thank you ;)

**Comment by zeffii** [comment]
the random number generator has many features, perhaps dig into it. Unfortunately it doesn't have a "custom function" feature. but personally I would use ScriptNode for that.. or indeed if the values are static write out a csv file once.

as for the .stl files.. I recall writing an ascii stl file loader at some point, but i'm failing to find it at the minute. Blender does have an .stl reader that also handles binary formats. It would take some poking around to implement that as a node directly.   There's nothing ready-made that i can point you to.

**Comment by javismiles** [comment]
thank you @zeffii , scriptNode is another sverchok node for basically writing python script to create new values?

**Comment by zeffii** [comment]
yep. we've included several useful templates and documentation, the issue tracker is also full of `snlite` "script not lite" snippets.

**Comment by javismiles** [comment]
thats great, will check it out ;)

**Comment by zeffii** [comment]
![image](https://user-images.githubusercontent.com/619340/69918568-015f3500-1474-11ea-86f1-971abc87867f.png)


```python
"""
in numpoints s d=200 n=2
out verts v
"""

import numpy as np
import math

X1 = np.linspace(0.0, 3*math.pi, numpoints)
Y1 = np.sin(X1)
Z1 = np.zeros(numpoints)
C1 = np.vstack([X1, Y1, Z1]).T.tolist()

verts.append(C1)
```

**Comment by zeffii** [comment]
@javismiles  please start new / unrelated questions as a new issue in the tracker, you can always reference this issue using `#2417` if you need. It's just nicer to not have huge threads whenever possible

**Comment by javismiles** [comment]
@zeffii you are totally right, I will stop this thread here and will open new ones, thank you again for all your amazing support ;)
